### ISSUE TITLE
Pairwise metric in LTR that take into account all the rows of each query 

### ISSUE THREAD

**Issue by blindape2** [microsoft/LightGBM#4755]
Is there a way to do LTR with pairwise metric, without any NDCG or label_gain setting?
Every reference I find is for problems in which the interest is to rank well the top of each queries.
How to deal with a problem in which we are interested in right sorting of all the query?
(By example, in a problem in which the original metric is spearman correlation of each query).

**Comment by ostrokach** [comment]
I have the exact same question!

I think a similar question was also asked previously (microsoft/LightGBM#877) but I fail to grok the answer.

XGBoost has `rank:pairwise` and CatBoost has [`PairLogit`](https://catboost.ai/en/docs/concepts/loss-functions-ranking#PairLogit) and [`LightGBM`](https://catboost.ai/en/docs/concepts/loss-functions-ranking#PairLogitPairwise), but I can't find an equivalent in LightGBM?

**Comment by guolinke** [comment]
very sorry, this objective function is not yet supported in LightGBM. 
you can achieve that by update `GetGradientsForOneQuery`. The simplest solution is to calculate gradients/hessians for all pairs, then sum them to rows. 

https://github.com/microsoft/LightGBM/blob/01568cf59a412de7823b373fe13eb36db9cffe53/src/objective/rank_objective.hpp#L140
